### DISCUSSION TITLE
scripts/tmuxinator.sh prerequisites and issues

### DISCUSSION THREAD

**Discussion by yojoe** [fedimint/fedimint#514] [Q&A] [answered]
I tried to walk through the tutorial [docs/dev-env.md](docs/dev-env.md) and [docs/dev-running.md](docs/dev-running.md) but ran into multiple issues/blockers and always got stuck at the "Setting up lightning channel" step.

Here's a list of issues I have:

#### 1. Prerequisites
Unfortunately, `Rust toolchain` and `Nix package manager` is not enough to successfully run [scripts/tmuxinator.sh](scripts/tmuxinator.sh) and follow the tutorial. You also need the following binaries in your `$PATH`, which isn't mentioned in the current docs:
- `bitcoind`
- `lightningd` which was compiled with `--enable-developer`
- `jq`
- `bc`

The last two ones (`jq` and `bc`) were not so easy to figure out, because the script-functions invoking them are called with `> /dev/null 2>&1`, so it took a lot of time debugging what's wrong. At a minimum, it would be helpful, if [docs/dev-running.md](docs/dev-running.md) would mention all the prerequisites.

However, what's weird is, that during the build phase all these requirements are downloaded/compiled and available within the `nix develop` shell, but not anymore in the shells [`tmuxinator.sh`](scripts/tmuxinator.sh) starts. [`tmuxinator.sh`](scripts/tmuxinator.sh) seems to launch "fresh" `bash` shells, where those binaries from the build phase a not in the `$PATH` anymore. So even if you run [scripts/tmuxinator.sh](scripts/tmuxinator.sh) from within a `nix develop` shell, you get tons of issues/errors due to the missing requirements. I spent a couple of hours investigating `nix` and `tmuxinator` how to change this behavior and pass through the whole `nix develop` environment to the `tmux` instances, but didn't manage to get it running. Maybe someone else could look into this.

In the end, I `hardcoded` the `nix-store` paths of the above prerequisites in the `pre_window` section of my [.tmuxinator.yml](.tmuxinator.yml), to get the demo/tutorial running without issues. I fetched the nix-store paths using `which` from within a `nix develop` shell. Of course this ugly and not resilient to changes, but at least I can run the tutorial now and it's just a single change to the upstream code base.
```
    1 name: fedimint-dev
    2 root: .
    3 socket_name: fedimint-dev
    4 pre_window:
    5   - source .tmpenv
    6   - alias ln1="\$FM_LN1"
    7   - alias ln2="\$FM_LN2"
    8   - alias btc_client="\$FM_BTC_CLIENT"
    9   - alias mint-client-cli="\$FM_MINT_CLIENT"
   10   - alias mint_rpc_client="\$FM_MINT_RPC_CLIENT"
+  11   - export PATH="/nix/store/djpgskxn580p9bjgnw6skv4knzg3i8rz-bitcoind-23.0/bin/:\
+  12                  /nix/store/7qrbkcabl319syqvchbfq0x7z1b5lbqy-clightning-0.11.1/bin/:\
+  13                  /nix/store/y98cnvi73q404kdb0slbddk5abv8g5bw-jq-1.6-bin/bin/:\
+  14                  /nix/store/g5wzdl003a65hgdqzgmaab4vqix035x0-bc-1.07.1/bin/:\
+  15                  $PATH"
   16 tmux_detached: false
   17 windows:
   ...
```

Maybe somebody finds a more elegant way to solve this. Or at least newbies looking into this project and trying to run the tutorial find my findings helpful to get going.

**Comment by justinmoon** [comment]
`tmuxinator.sh` should run itself inside a nix shell which has all the dependencies. Could you post the error message you get when you try to just run `./scripts/tmuxinator.sh`? Also, can you share any details about your system like OS etc?

**Comment by yojoe** [comment]
> Also, can you share any details about your system like OS etc?

Debian 11 using `zsh` instead of `bash` and invoking `nix develop -c zsh` within the project root first, before launching [scripts/tmuxinator.sh](scripts/tmuxinator.sh).

> Could you post the error message you get when you try to just run ./scripts/tmuxinator.sh?

The only error I could spot as a complete newcomer to the project was `command not found: lightningd` in the two tmux lightning panes on the right after switching to the second window, else you just get the impression it's still waiting for `Setting up lightning channel` step to finish. You don't see all the errors that happened before because `jq` and `bc` are also not available by default on Debian 11.

**Comment by yojoe** [comment]
> tmuxinator.sh should run itself inside a nix shell which has all the dependencies.

When I type in the shell of the first pane (`user`) of the second tmux window (`federation`):
```sh
~/Projects/fedimint$ echo $IN_NIX_SHELL
impure
~/Projects/fedimint$ echo $SHELL
/nix/store/bmwq0pq3xbzvb3l95zpmwcg05mzisjr2-bash-interactive-5.1-p16/bin/bash
~/Projects/fedimint$ which lightningd
~/Projects/fedimint$ which jq
~/Projects/fedimint$ which bc
~/Projects/fedimint$ 
~/Projects/fedimint$ echo $PATH
/home/yojoe/.local/bin:/home/yojoe/bin:/home/yojoe/.cargo/bin:/home/yojoe/.nix-profile/bin:/nix/var/nix/profiles/default/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

But when I type `env` there's a huge amount of nix stuff in it, so yes, it's probably a `nix shell` but for some reason only `$PATH` is not filled with all the `nix-store` paths.

**Comment by dpc** [comment]
> Unfortunately, `Rust toolchain` and `Nix package manager` is not enough to successfully run [scripts/tmuxinator.sh](/scripts/tmuxinator.sh) and follow the tutorial. You also need the following binaries in your `$PATH`, w

All the commands need to be executed in a terminal with `nix develop` shell started. To check if that's the case, you can:

```
dpc@futex:/home/dpc/lab/fedimint$ echo $IN_NIX_SHELL 
impure
```

It is possible to work on fedimint without it, but indeed the developer will have to take care of setting up all these things themselves.


```
> nix develop -i                                           
dpc@futex:/home/dpc/lab/fedimint$ echo $IN_NIX_SHELL 
impure
dpc@futex:/home/dpc/lab/fedimint$ which bitcoind 
/nix/store/1af4cixag6kzpin0iajx09aphkp2snx8-bitcoind-23.0/bin/bitcoind        
dpc@futex:/home/dpc/lab/fedimint$ which lightningd
/nix/store/xyrdsgamphh607x23srw6dgks5884yx7-clightning-0.11.1/bin/lightningd
dpc@futex:/home/dpc/lab/fedimint$ which bc
/nix/store/rkcjmvvqmxb91cpb7an4amjsnj5qq5bw-bc-1.07.1/bin/bc
dpc@futex:/home/dpc/lab/fedimint$ which jq
/nix/store/6b4xh3911c0gdiwkfbiirafra1mblvw9-jq-1.6-bin/bin/jq
```

**Comment by justinmoon** [comment]
BTW we might be able to help you debug this quicker in [our discord](https://discord.gg/dZYajBMsEB)

**Comment by yojoe** [comment]
> All the commands need to be executed in a terminal with `nix develop` shell started.

Yes, the `nix develop` shell/terminal is fine on my end, all the needed binaries are in the $PATH and can be found using `which`.  Only the shells (panes) created by [scripts/tmuxinator.sh](scripts/tmuxinator.sh) don't have the $PATH setup correctly.

<img width="1512" alt="Screenshot 2022-09-07 at 20 34 51" src="https://user-images.githubusercontent.com/58518123/188954276-08d8438e-e71f-4313-a19c-563b2ce8a3c8.png">

**Comment by dpc** [comment]
I just tried on my system (NixOS) and all the panes start the binaries just fine.

Something on your system is preventing the shells inside tmux from inheriting the environment from the outside. Either the session is already maybe opened (try ctrl+b `:kill-session` or something like that), or possibly your bash scripts reset the env, or there's some tmux config.

**Comment by dpc** [comment]
A workaround would be to prefix commands inside `.tmuxinator.yml` with `nix develop -c ...`

**Comment by dpc** [comment]
Something like this:

```
> git diff HEAD
diff --git a/.tmuxinator.yml b/.tmuxinator.yml
index 8e9f37ec0..735a2071a 100644
--- a/.tmuxinator.yml
+++ b/.tmuxinator.yml
@@ -20,20 +20,20 @@ windows:
         - user:
           - # empty user shell
         - bitcoind:
-          - bitcoind -regtest -fallbackfee=0.0004 -txindex -server -rpcuser=bitcoin -rpcpassword=bitcoin -datadir=$FM_BTC_DIR &
+          - nix develop -c bitcoind -regtest -fallbackfee=0.0004 -txindex -server -rpcuser=bitcoin -rpcpassword=bitcoin -datadir=$FM_BTC_DIR &
           - echo $! >> $FM_PID_FILE
           - fg
         - ln1:
           - sleep 1 # wait for bitcoind
-          - lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN1_DIR --addr=127.0.0.1:9000 --plugin=$FM_BIN_DIR/ln_gateway --fedimint-cfg=$FM_CFG_DIR &
+          - nix develop -c lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN1_DIR --addr=127.0.0.1:9000 --plugin=$FM_BIN_DIR/ln_gateway --fedimint-cfg=$FM_CFG_DIR &
           - echo $! >> $FM_PID_FILE
           - fg
         - ln2:
           - sleep 1 # wait for bitcoind
-          - lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN2_DIR --addr=127.0.0.1:9001 &
+          - nix develop -c lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN2_DIR --addr=127.0.0.1:9001 &
           - echo $! >> $FM_PID_FILE
           - fg
         - federation:
           - sleep 1 # wait for bitcoind
-          - ./scripts/start-fed.sh
+          - nix develop -c ./scripts/start-fed.sh
```

kind of works for me.

**Comment by yojoe** [comment, marked as answer]
Found the root cause. This thread gave a clue: https://superuser.com/questions/544989/does-tmux-sort-the-path-variable

> You can't easily persuade tmux (or rather, bash) not to source /etc/profile (for some reason tmux always runs as a login shell, which means /etc/profile will be read)...

At least on Debian 11 `/etc/profile` sets the $PATH variable:

```sh
cat /etc/profile
# /etc/profile: system-wide .profile file for the Bourne shell (sh(1))
# and Bourne compatible shells (bash(1), ksh(1), ash(1), ...).

if [ "$(id -u)" -eq 0 ]; then
  PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
else
  PATH="/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games"
fi
export PATH
...
```
This means, whenever a `tmux` instance is started `/etc/profile` is sourced **again** and the `$PATH` which was previously set by `nix develop` is overridden again.

IMHO, the best solution is from one of the comments of the accepted answer:

> If you don't want to mess with system defaults, `set -g default-command "${SHELL}"` in `.tmux.conf` forces tmux to use non-login shells. I don't see the point in those anyway, since you usually open tmux after you've already logged in. – seeker_of_bacon Jan 18, 2020 at 16:56

That's exactly what I did and created the following `~/.tmux.conf` in my home directory:
```
❯ cat ~/.tmux.conf
set -g default-command "${SHELL}"
```

Now everything is fine and running [scripts/tmuxinator.sh](scripts/tmuxinator.sh) works fine without any issues! I'll open a PR to mention this in [docs/dev-env.md](docs/dev-env.md).
